### PR TITLE
feat: setup canary releases using changesets

### LINKED ISSUE #3383: Canary releases
As discussed on graphql-js-wg we decided to switch to unstable `main` and provide canary release every week.

Currently, there are a couple of blockers we need to address:
- [ ] Allow publishing NPM packages without updating the `main` (otherwise we will add a bunch of useless commits in git history).
   - [ ] Don't commit `graphql/version.ts` and only generate it during the build process.
   - [ ] Figure out a way to safely release versions without updating `package.json`
- [x] Setup docs site for v16 we should have docs for "stable" version.
- [x] Release v16.1.0
- [x] Switch main to v17.0.0-alpha.0

cc: @saihaj

### PR DESCRIPTION
**Pull request by saihaj**
Need to setup GitHub secret for NPM token `NODE_AUTH_TOKEN` and Bot token to comment `GH_API_TOKEN`

closes #3383

### CODE REVIEW

File: .github/workflows/canary.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      - name: Publish a message
+        if: steps.canary.outputs.released
+        uses: 'kamilkisiela/pr-comment@master'
+        with:
+          message: |
+            The latest changes of this PR are available as alpha in npm (based on the declared `changesets`):
+            ```
+            ${{ steps.canary.outputs.changesetsPublishedPackages}}
+            ```
+          bot-token: ${{ secrets.GH_API_TOKEN }}
```
Comment (dotansimha) [inline .github/workflows/canary.yml:52]: you can use `${{ secrets.GITHUB_TOKEN }}` here, check the setup in `graphiql` repo

File: .github/workflows/canary.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+          bot: 'theguild-bot'
```
Comment (dotansimha) [inline .github/workflows/canary.yml:53]: i think you can use `github-actions`

File: .github/workflows/canary.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
+name: Canary Release
```
Comment (dotansimha) [inline .github/workflows/canary.yml:1]: I guess we need the 2nd workflow for release? this way we can release using `changeset` as well. What do you think @IvanGoncharov ? 
(Changesets collects all changeset files and keep an updated Release PR, so when it's merged, you get the actual version bump and NPM release)

File: resources/canary-release.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
+/* eslint-disable */
```
Comment (dotansimha) [inline resources/canary-release.js:1]: This is optional, we are using that to do custom pre-release and not use the default way of doing that with `changesets`...

File: .github/workflows/canary.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        id: canary
+        uses: 'kamilkisiela/release-canary@master'
+        with:
+          npm-token: ${{ secrets.NODE_AUTH_TOKEN }}
+          npm-script: 'npm run release:canary'
+          changesets: true
+
+      - name: Publish a message
+        if: steps.canary.outputs.released
+        uses: 'kamilkisiela/pr-comment@master'
```
Comment (IvanGoncharov) [inline .github/workflows/canary.yml:45]: This custom guild stuff, right?
I think idea was that we use some popular tool for doing canary releases?
Those look like there are guild internal scripts?
I'm ok with automation but it should be something popular and widespread.

**Review by IvanGoncharov — CHANGES_REQUESTED** [review]

File: .github/workflows/canary.yml (hunk 5 above)
Comment (kamilkisiela) [reply .github/workflows/canary.yml:45]: I can make it `facebook/pr-comment`, does it change anything?

Comment (kamilkisiela) [reply .github/workflows/canary.yml:45]: how it's different from https://github.com/graphql/graphql-js/blob/main/resources/build-deno.js ?

File: .github/workflows/canary.yml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        uses: actions/setup-node@v2
+        with:
+          cache: npm
+          node-version: ${{ env.NODE_VERSION_USED_FOR_DEVELOPMENT }}
+
+      - name: Install Dependencies
+        run: npm ci --ignore-scripts
+
+      - name: Setup NPM credentials
+        run: echo "//registry.npmjs.org/:_authToken=$NODE_AUTH_TOKEN" >> ~/.npmrc
```
Comment (IvanGoncharov) [inline .github/workflows/canary.yml:31]: I'm not ok with putting NPM token on FS, it means all subsequent scripts have access to it.
Also, I'm not ok passing NPM token to anything that is not official NPM or GitHub tool/lib/etc.

File: .github/workflows/canary.yml (hunk 3 above)
Comment (IvanGoncharov) [reply .github/workflows/canary.yml:1]: We discuss canary release so let's address that and see how it goes.

**Pull request closed without merging**